src/main/java/com/salesforce/dataloader/controller/Controller.java

```diff
@@ -234,43 +241,87 @@ public synchronized boolean saveConfig() {
+     *  False if directory was failed to create
      */
-    private void copyFile(File file, File destFile) {
-        BufferedInputStream in = null;
-        BufferedOutputStream out = null;
-        try {
-
-            // reading and writing 8 KB at a time
-            int bufferSize = 8 * 1024;
-
-            byte[] buf = new byte[bufferSize];
+    private static boolean createDir(File dirPath) {
+        boolean isSuccessful = true;
+        if (!dirPath.exists() || !dirPath.isDirectory()) {
+            isSuccessful = dirPath.mkdirs();
+            if (isSuccessful) {
+                logger.info("Created config directory: " + dirPath);
+            } else {
+                logger.info("Unable to create config directory: " + dirPath);
+            }
+        } else {
+            logger.info("Config directory already exists: " + dirPath);
+        }
+        return isSuccessful;
+    }
 
-            in = new BufferedInputStream(new FileInputStream(file), bufferSize);
-            int r = 0;
+    /**
+     * Returns current user's Dataloader configuration directory
+     *  ie) For Windows - C:\Users\{user}\AppData\Local\salesforce.com\Data Loader {version}\conf
+     *      For Mac - /Users/{user}/Library/Preferences/salesforce.com/Data Loader {version}/conf
+     *      For *nix - ~/.salesforce.com/Data Loader {version}/conf
+     *
+     * @return
+     *  Current user's Dataloader configuration directory
+     */
+    private static File getUserConfigDir() {
+        File dir;
+        switch (OS_TYPE) {
+            case WINDOWS: {
+                dir = Paths.get(System.getProperty("user.home"), "AppData\\Local", APP_VENDOR, getProductName(), CONFIG_DIR).toFile();
+                break;
+            }
+            case MACOSX: {
+                dir = Paths.get(System.getProperty("user.home"), "Library/Preferences", APP_VENDOR, getProductName(), CONFIG_DIR).toFile();
+                break;
+            }
+            default:
+            case LINUX: {
+                dir = Paths.get(System.getProperty("user.home"), "." + APP_VENDOR, getProductName(), CONFIG_DIR).toFile();
+                break;
+            }
 
-            out = new BufferedOutputStream(new FileOutputStream(destFile), bufferSize);
+        }
+        return dir;
+    }
 
-            while ((r = in.read(buf, 0, buf.length)) != -1) {
-                out.write(buf, 0, r);
+    /**
+     * Returns default Dataloader configuration directory
+     * ie) For Windows - C:\Program Files (x86)\salesforce.com\Data Loader\conf
+     *     For Mac - /Applications/Data Loader.app/Contents/Resources/conf
+     *     For *nix - {where DL is installed}/conf
+     *
+     * @return
+     *  Default Dataloader configuration directory
+     */
+    private static File getDefaultConfigDir() {
+        File dir;
+        switch (OS_TYPE) {
+            case WINDOWS: {
+                dir = new File(System.getProperty("user.dir"), "conf");
```

Comment:
The intention is used to get the configuration under user directory.

Reply:
Pretty sure that's incorrect. You're thinking of `getUserConfigDir` (see Controller.java:292); the `getDefaultConfigDir` function is supposed to return the dir containing the default config template, which is under the user directory (see pom.xml:771-773 and associated comment, setup.nsi:195-196,313,317, and Controller.java:88,375-390 for usage).

Again, this isn't a security bug, it just won't (reliably) work the way it's supposed to. Also, the new comment is still misleading: `System.getProperty("user.dir")` returns the working directory, which is (usually) not the user's (home) directory; it might be the install location (which is the only case under which this code will work), the user's DataLoader config directory (as set by the Start menu shortcut on Windows, see setup.nsi:7,201), or wherever the user's shell has as its current working directory (could be anywhere).

Reply:
Agree that was wrong.  Will test: String path = Controller.class.getProtectionDomain()
                .getCodeSource().getLocation().getPath();   and verify if it works.